src/main/java/duke/command/DoneCommand.java

```diff
@@ -0,0 +1,42 @@
+package duke.command;
+
+import duke.DukeException;
+import duke.storage.Storage;
+import duke.task.TaskList;
+import duke.ui.Ui;
+
+/**
+ * Marks task as done.
+ */
+public class DoneCommand extends Command {
+
+    private int taskNo;
+
+    /**
+     * Constructs a DoneCommand object.
+     *
+     * @param taskNo the index of the task that is completed.
+     */
+    public DoneCommand(int taskNo) {
+        super(false);
+        this.taskNo = taskNo;
+    }
+
+    /**
+     * Executes DoneCommand by marking the task as done.
+     *
+     * @param tasks   TaskList of Duke.
+     * @param ui      The user interface.
+     * @param storage To load from and save to the disk.
+     * @return Acknowledgement message sent by Duke.
+     * @throws DukeException if the given task number is out of bound of the task list.
+     */
+    @Override
+    public String execute(TaskList tasks, Ui ui, Storage storage) throws DukeException {
+        try {
+            return tasks.getTasks().get(taskNo).markAsDone();
```

Comment:
Related to the point about copying the list rather than returning it, it might be better to let the `TaskList` to mark the task as done. This ensures that the command cannot touch the actual task itself, but deals with it via `TaskList`.